package/MDAnalysis/lib/formats/include/readdcd.h

```diff
@@ -0,0 +1,790 @@
+    rc = fio_fseek(fd, sizeof(float)*(N-upperb-1)+sizeof(int)*2+sizeof(float)*lowerb, FIO_SEEK_CUR); /* skip 2 format integers */
+    if (rc == -1) return DCD_BADREAD;
+		
+    if (fio_fread(Y, sizeof(float)*range, 1, fd) != 1) return DCD_BADREAD; 						 /* read Y coordinates */
+    rc = fio_fseek(fd, sizeof(float)*(N-upperb-1)+sizeof(int)*2+sizeof(float)*lowerb, FIO_SEEK_CUR); /* skip 2 format integers */
+    if (rc == -1) return DCD_BADREAD;
+		
+    if (fio_fread(Z, sizeof(float)*range, 1, fd) != 1) return DCD_BADREAD; 						 /* read Z coordinates */
+    rc = fio_fseek(fd, sizeof(float)*(N-upperb-1)+sizeof(int), FIO_SEEK_CUR);   			 /* skip 1 format integer */
+    if (rc == -1) return DCD_BADREAD;
+		
+    /* convert endianism if necessary */
+    if (reverseEndian) {
+      swap4_aligned(X, range);
+      swap4_aligned(Y, range);
+      swap4_aligned(Z, range);
+    }
+
+    /* copy fixed atom coordinates into fixedcoords array if this was the */
+    /* first timestep, to be used from now on.  We just copy all atoms.   */
+    /*if (num_fixed && first) {
+      memcpy(fixedcoords, X, range*sizeof(float));
+      memcpy(fixedcoords+range, Y, range*sizeof(float));
+      memcpy(fixedcoords+2*range, Z, range*sizeof(float));
+      }*/
+
+    /* skip the optional charmm 4th array */
+    /* XXX this too should be read together with the other items in a */
+    /*     single fio_readv() call in order to prevent lots of extra  */
+    /*     kernel/user context switches.                              */
+    if ((charmm & DCD_IS_CHARMM) && (charmm & DCD_HAS_4DIMS)) {
+      if (fio_fread(&input_integer, sizeof(int), 1, fd) != 1) return DCD_BADREAD;
+      if (reverseEndian) swap4_aligned(&input_integer, 1);
+      if (fio_fseek(fd, input_integer+sizeof(int), FIO_SEEK_CUR)) return DCD_BADREAD;
+    }
+    //ret_val = read_charmm_4dim(fd, charmm, reverseEndian);
+    //if (ret_val) return ret_val;
+  } else {
+    return DCD_BADFORMAT;
+  }
+  if (unitcell[1] >= -1.0 && unitcell[1] <= 1.0 &&
+      unitcell[3] >= -1.0 && unitcell[3] <= 1.0 &&
+      unitcell[4] >= -1.0 && unitcell[4] <= 1.0) {
+    /* This file was generated by Charmm, or by NAMD > 2.5, with the angle */
+    /* cosines of the periodic cell angles written to the DCD file.        */
+    /* This formulation improves rounding behavior for orthogonal cells    */
+    /* so that the angles end up at precisely 90 degrees, unlike acos().   */
+    /* (changed in MDAnalysis 0.9.0 to have NAMD ordering of the angles;   */
+    /* see Issue 187) */
+    alpha = 90.0 - asin(unitcell[4]) * 90.0 / M_PI_2;
+    beta  = 90.0 - asin(unitcell[3]) * 90.0 / M_PI_2;
+    gamma = 90.0 - asin(unitcell[1]) * 90.0 / M_PI_2;
+  } else {
+    /* This file was likely generated by NAMD 2.5 and the periodic cell    */
+    /* angles are specified in degrees rather than angle cosines.          */
+    alpha = unitcell[4];
+    beta  = unitcell[3];
+    gamma = unitcell[1];
+  }
```

Comment:
We could simply read out the 6 numbers and then have the reader figure it out in the beginning and then use one conversion function in the layer that converts to/from dimensions. This would make the C/pyx code fairly clean: we really just read what's there.

I don't know how to autodetect without actually reading the first frame, though.

I also don't know if there's a significant speed penalty for moving the `asin()` etc out of the C layer into Python – probably not, compared to reading the rest, but one potential consideration.